### PR TITLE
feat!: allow consumer chains to change their PSS parameters

### PR DESCRIPTION
**Pull request by insumity**
## Description

This PR introduces the `ConsumerModificationProposal` that allows consumer chains to change PSS parameters (e.g., `topN`, `validatorSetCap`, etc.) while a consumer chain is running.

Note that a validator that was automatically opted in on a Top N consumer chain would remain opted in even after moving to an Opt In chain.

---

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

* [x] Included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
* [x] Added `!` to the type prefix if the change is [state-machine breaking](https://github.com/cosmos/interchain-security/blob/main/RELEASES.md#breaking-changes)
* [ ] Confirmed this PR does not introduce changes requiring state migrations, OR migration code has been added to consumer and/or provider modules
* [x] Targeted the correct branch (see [PR Targeting](https://github.com/cosmos/interchain-security/blob/main/CONTRIBUTING.md#pr-targeting))
* [ ] Provided a link to the relevant issue or specification
* [ ] Followed the guidelines for [building SDK modules](https://github.com/cosmos/cosmos-sdk/blob/main/docs/build/building-modules/00-intro.md)
* [ ] Included the necessary unit and integration [tests](https://github.com/cosmos/interchain-security/blob/main/CONTRIBUTING.md#testing)
* [ ] Added a changelog entry to `CHANGELOG.md`
* [ ] Included comments for [documenting Go code](https://blog.golang.org/godoc)
* [ ] Updated the relevant documentation or specification
* [ ] Reviewed "Files changed" and left comments if necessary 
* [ ] Confirmed all CI checks have passed
* [ ] If this PR is library API breaking, bump the go.mod version string of the repo, and follow through on a new major release

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

* [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
* [ ] confirmed `!` the type prefix if the change is state-machine breaking
* [ ] confirmed this PR does not introduce changes requiring state migrations, OR confirmed migration code has been added to consumer and/or provider modules
* [ ] confirmed all author checklist items have been addressed
* [ ] reviewed state machine logic
* [ ] reviewed API design and naming
* [ ] reviewed documentation is accurate
* [ ] reviewed tests and test coverage

## Summary by CodeRabbit

- **New Features**
  - Introduced a new governance proposal type, `ConsumerModificationProposal`, allowing modifications to parameters of a running consumer chain.
  - Added functionality to handle and submit consumer modification proposals, including JSON parsing and validation.

- **Bug Fixes**
  - Added error handling for invalid consumer modification proposals.

- **Tests**
  - Implemented comprehensive test cases for the new `ConsumerModificationProposal` to ensure proper validation and functionality.

- **Documentation**
  - Updated documentation to reflect new proposal type and related functionalities.

### CODE REVIEW

File: x/ccv/provider/client/proposal_handler.go [hunk 1]
```diff
@@ -172,6 +173,70 @@ Where proposal.json contains:
...
+The proposal details must be supplied via a JSON file.
+
+Example:
+$ <appd> tx gov submit-legacy-proposal consumer-modification <path/to/proposal.json> --from=<key_or_address>
+
+Where proposal.json contains:
+
+{
+    "title": "Modify FooChain",
+    "summary": "Make it an Opt IN chain",
```
Comment (p-offtermatt) [inline x/ccv/provider/client/proposal_handler.go:194]:
```suggestion
    "summary": "Make it an Opt In chain",
```
unless the capitalization was for emphasis?

File: x/ccv/provider/keeper/partial_set_security.go [hunk 2]
```diff
@@ -290,6 +290,7 @@ func (k Keeper) ComputeNextValidators(ctx sdk.Context, chainID string, bondedVal
 			return k.FilterOptedInAndAllowAndDenylistedPredicate(ctx, chainID, providerAddr)
 		})
 
+	// if ELYS ... to 50 ..
```
Comment (p-offtermatt) [inline x/ccv/provider/keeper/partial_set_security.go:293]: remove comment

File: x/ccv/provider/types/proposal.go [hunk 3]
```diff
@@ -92,6 +95,21 @@ func (cccp *ConsumerAdditionProposal) ProposalType() string {
...
+// ValidatePSSFeatures returns an error if the `topN` and `validatorsPowerCap` parameters are no in the correct ranges
+func ValidatePSSFeatures(topN uint32, validatorsPowerCap uint32, error *errorsmod.Error) error {
+	// Top N corresponds to the top N% of validators that have to validate the consumer chain and can only be 0 (for an
+	// Opt In chain) or in the range [50, 100] (for a Top N chain).
+	if topN != 0 && (topN < 50 || topN > 100) {
+		return errorsmod.Wrap(error, "Top N can either be 0 or in the range [50, 100]")
+	}
+
+	if validatorsPowerCap != 0 && validatorsPowerCap > 100 {
+		return errorsmod.Wrap(error, "validators' power cap has to be in the range [1, 100]")
```
Comment (p-offtermatt) [inline x/ccv/provider/types/proposal.go:107]:
```suggestion
		return errorsmod.Wrap(error, "validators' power cap must has to be in the range [0, 100]")
```

File: x/ccv/provider/types/proposal.go [hunk 4]
```diff
@@ -145,17 +163,7 @@ func (cccp *ConsumerAdditionProposal) ValidateBasic() error {
...
-	if cccp.Top_N != 0 && (cccp.Top_N < 50 || cccp.Top_N > 100) {
-		return errorsmod.Wrap(ErrInvalidConsumerAdditionProposal, "Top N can either be 0 or in the range [50, 100]")
-	}
-
-	if cccp.ValidatorsPowerCap != 0 && cccp.ValidatorSetCap > 100 {
-		return errorsmod.Wrap(ErrInvalidConsumerAdditionProposal, "validators' power cap has to be in the range [1, 100]")
-	}
-
-	return nil
+	return ValidatePSSFeatures(cccp.Top_N, cccp.ValidatorsPowerCap, ErrInvalidConsumerAdditionProposal)
```
Comment (sainoe) [inline x/ccv/provider/types/proposal.go:166]: nits: I'd rather handle the error here and wrap it with `ErrInvalidConsumerAdditionProposal`.

File: x/ccv/provider/types/proposal.go [hunk 5]
```diff
@@ -92,6 +95,21 @@ func (cccp *ConsumerAdditionProposal) ProposalType() string {
 	return ProposalTypeConsumerAddition
 }
 
+// ValidatePSSFeatures returns an error if the `topN` and `validatorsPowerCap` parameters are no in the correct ranges
+func ValidatePSSFeatures(topN uint32, validatorsPowerCap uint32, error *errorsmod.Error) error {
```
Comment (sainoe) [inline x/ccv/provider/types/proposal.go:99]: nits: wrapping the returned error of `ValidatePSSFeatures` is cleaner than passing it as an argument.

File: x/ccv/provider/types/proposal.go [hunk 6]
```diff
@@ -223,6 +231,47 @@ func (sccp *ConsumerRemovalProposal) ValidateBasic() error {
...
+func (cccp *ConsumerModificationProposal) ValidateBasic() error {
+	if err := govv1beta1.ValidateAbstract(cccp); err != nil {
+		return err
+	}
+
+	if strings.TrimSpace(cccp.ChainId) == "" {
+		return errorsmod.Wrap(ErrInvalidConsumerModificationProposal, "consumer chain id must not be blank")
+	}
+
+	return ValidatePSSFeatures(cccp.Top_N, cccp.ValidatorsPowerCap, ErrInvalidConsumerModificationProposal)
```
Comment (sainoe) [inline x/ccv/provider/types/proposal.go:272]: See comment above.

**Comment by sainoe** [comment]
LGTM! Just one nit.

**Review by sainoe — APPROVED** [review]

**Review by p-offtermatt — APPROVED** [review]

File: x/ccv/provider/types/proposal.go (hunk 6 above)
Comment (insumity) [reply x/ccv/provider/types/proposal.go]: Fixed.

File: x/ccv/provider/types/proposal.go (hunk 5 above)
Comment (insumity) [reply x/ccv/provider/types/proposal.go]: Yes, that's much cleaner. Thanks for the suggestion.

**Pull request merged**